source/fuzz/transformation_changing_memory_semantics.cpp

```diff
@@ -0,0 +1,194 @@
+  *message_.mutable_atomic_instruction() = atomic_instruction;
+  message_.set_memory_semantics_operand_index(memory_semantics_operand_index);
+  message_.set_memory_semantics_new_value_id(memory_semantics_new_value_id);
+}
+
+bool TransformationChangingMemorySemantics::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  auto needed_atomic_instruction =
+      FindInstruction(message_.atomic_instruction(), ir_context);
+  // The atomic instruction must exist.
+  if (!needed_atomic_instruction) {
+    return false;
+  }
+
+  // Memory semantics operand Index must be equal to 2 or 3 only.
+  auto operand_index = message_.memory_semantics_operand_index();
+  if (operand_index != 2 && operand_index != 3) {
+    return false;
+  }
+
+  // The new instruction value must exist and must be OpConstant.
+  auto new_instruction = ir_context->get_def_use_mgr()->GetDef(
+      message_.memory_semantics_new_value_id());
+  if (!new_instruction) {
+    return false;
+  }
+  if (new_instruction->opcode() != SpvOpConstant) {
+    return false;
+  }
+  auto new_memory_sematics_value = static_cast<SpvMemorySemanticsMask>(
+      new_instruction->GetSingleWordInOperand(0));
+
+  // Instruction must be atomic instruction only. Operand index must be equal to
+  // 2 in the case of instruction that takes one memory semantics operand.
+  // Instructions that takes two memory semantics, one of the value must be
+  // stronger than the other.
+  switch (needed_atomic_instruction->opcode()) {
+    case SpvOpAtomicLoad:
+    case SpvOpAtomicStore:
+    case SpvOpAtomicExchange:
+    case SpvOpAtomicIIncrement:
+    case SpvOpAtomicIDecrement:
+    case SpvOpAtomicIAdd:
+    case SpvOpAtomicISub:
+    case SpvOpAtomicSMin:
+    case SpvOpAtomicUMin:
+    case SpvOpAtomicSMax:
+    case SpvOpAtomicUMax:
+    case SpvOpAtomicAnd:
+    case SpvOpAtomicOr:
+    case SpvOpAtomicXor:
+    case SpvOpAtomicFlagTestAndSet:
+    case SpvOpAtomicFlagClear:
+    case SpvOpAtomicFAddEXT:
+      if (operand_index != 2) {
+        return false;
+      }
+      break;
+
+    case SpvOpAtomicCompareExchange:
+    case SpvOpAtomicCompareExchangeWeak:
+      if (new_memory_sematics_value ==
+          static_cast<SpvMemorySemanticsMask>(
+              ir_context->get_def_use_mgr()
+                  ->GetDef(needed_atomic_instruction->GetSingleWordInOperand(
+                      operand_index == 3 ? 2 : 3))
+                  ->GetSingleWordInOperand(0))) {
+        return false;
+      }
+      break;
+
+    default:
+      return false;
+  }
+
+  // The new memory semantics value must be larger than old.
+  auto old_memory_sematics_value = static_cast<SpvMemorySemanticsMask>(
+      ir_context->get_def_use_mgr()
+          ->GetDef(
+              needed_atomic_instruction->GetSingleWordInOperand(operand_index))
+          ->GetSingleWordInOperand(0));
+  if (new_memory_sematics_value <= old_memory_sematics_value) {
+    return false;
+  }
```

Comment:
I think we should remove this check. We actually need to do a more restrictive check, so this is potentially just confusing. Once we get to the point where we think the new value is OK, we could _then_ assert that it is larger than the old value, as a safety check. But we have more checks to do first. 

I think you will end up having a similar check in `IsValidConverstion`, though.